src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs

```diff
@@ -592,56 +592,60 @@ private void UpdateTextBoxes(Configuration.PackageSource packageSource)
             }
         }
 
-        [SuppressMessage("Microsoft.VisualStudio.Threading.Analyzers", "VSTHRD010", Justification = "NuGet/Home#4833 Baseline")]
         private void OnBrowseButtonClicked(object sender, EventArgs e)
         {
             const int MaxDirectoryLength = 1000;
 
             //const int BIF_RETURNONLYFSDIRS = 0x00000001;   // For finding a folder to start document searching.
             const int BIF_BROWSEINCLUDEURLS = 0x00000080; // Allow URLs to be displayed or entered.
 
-            var uiShell = (IVsUIShell2)_serviceProvider.GetService(typeof(SVsUIShell));
-
-            char[] rgch = new char[MaxDirectoryLength + 1];
-
-            // allocate a buffer in unmanaged memory for file name (string)
-            IntPtr bufferPtr = Marshal.AllocCoTaskMem((rgch.Length + 1) * 2);
-            // copy initial path to bufferPtr
-            Marshal.Copy(rgch, 0, bufferPtr, rgch.Length);
-
-            VSBROWSEINFOW[] pBrowse = new VSBROWSEINFOW[1];
-            pBrowse[0] = new VSBROWSEINFOW
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
             {
-                lStructSize = (uint)Marshal.SizeOf(pBrowse[0]),
-                dwFlags = BIF_BROWSEINCLUDEURLS,
-                pwzDlgTitle = Resources.BrowseFolderDialogDescription,
-                nMaxDirName = MaxDirectoryLength,
-                hwndOwner = Handle,
-                pwzDirName = bufferPtr,
-                pwzInitialDir = DetermineInitialDirectory()
-            };
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
```

Comment:
Click events seem like a very reasonable place to call `ThreadHelper.ThrowIfNotOnUIThread()` instead of using JTF.Run and a delegate.